_posts/2024-08-19-given-priority-change-3.7.md

```diff
@@ -0,0 +1,232 @@
+---
+layout: blog-detail
+post-type: blog
+by: Oliver Bračevac, EPFL
+title: "Upcoming Changes to Givens in Scala 3.7"
+---
+
+## New Prioritization of Givens in Scala 3.7
+
+Scala 3.7 introduces changes to how `givens` are resolved, which can
+affect program behavior when multiple givens are present.
+
+For example, consider a library that provides a default
+`given` for a component:
+```scala
+// library code
+class LibComponent:
+    def msg = "library-defined"
+
+// default provided by library
+given libComponent: LibComponent = LibComponent()
+
+def printComponent(using c:LibComponent) = println(c.msg)
+```
+
+Up until Scala 3.6, clients of the library could override
+`libComponent` with a user-defined one through subtyping
+
+```scala
+// client code
+class UserComponent extends LibComponent:
+    override def msg = "user-defined"
+
+given userComponent: UserComponent = UserComponent()
+
+@main def run = printComponent
+```
+
+Let's run the example:
+
+```scala
+run // Scala <= 3.6: prints "user-defined"
+    // Scala 3.7: prints "library-defined"
+```
+
+What happened? In Scala 3.6 and earlier, the compiler prioritizes the
+`give`n with the _most specific_ compatible subtype
+(`userComponent`). However, in Scala 3.7, it selects the value with the
+_most general_ subtype instead (`libComponent`).
+
+
+### Motivation: Better Handling of Inheritance Triangles & Typeclasses
```

Comment:
We might switch the order of `Motivation` and `New Prioritization of Givens`. 
By doing so we're starting with the overall language improvement (positive vibes), then pointing out the possible drawbacks (negativity), followed by migration guides (neutral, slightly positive).
So the overall structure might look like: 
```
# New Prioritization of Givens in Scala 3.7 (intro)

Scala 3.7 will introduce changes to how `givens` are resolved, which can
affect program behaviour when multiple givens are present. 
The goal of this change is to make a givens resolution more predictable, however, it can lead to problems when migrating to Scala 3.7 or later. 
In this article we'd like to explain the motivation, possible issues and migration guides to prepare developers for the upcoming changes.
...

### Motivation 
Why change the priority to the `given` .... 

### Source incompatibility of the new givens prioritization 
Unfortunately, this change might affect source compatibility of some Scala codebases. Let's consider a library that provides a default
... <here example from the 1st paragraph>

```

Reply:
Done!